CIP-????/README.md

```diff
@@ -0,0 +1,471 @@
+3. Creating the script context would require the ledger to take a Haskell value and lift it into a Plutus Core term, which as we've seen is difficult.
+
+The solution that we chose was to pick a single generic structured data type which would be used for data interchange between the ledger and scripts.
+This is the 'Data' type.
+
+However, Data is not the natural representation of structured values inside a script, that would be as datatypes.
+So this means we often want a decoding step where the script translates Data into its own representation.
+
+### Problems With the Status Quo
+
+There are a few problems with the status quo.
+
+#### 1. Decoding the Script Context is Too Expensive
+
+As we progressed to more realistic (or indeed, real) examples, two things became clear:
+1. Real-world script contexts can be large, especially due to the complexity of values stored in outputs
+2. The cost of processing the script context is too high in any case
+
+These combine to make decoding real world script contexts prohibitive in many cases.
+Budget profiling of our examples in `plutus-use-cases` revealed that many were spending 30-40% of their budget just on decoding the script context, which is unacceptable.
+
+In reality, this is well known among developers, and most have moved to _not_ decoding the script context, and instead working directly with the underlying Data object.
+This is much more error prone, but at least provides a workaround.
+
+#### 2. Everything Can Always be Faster
+
+The Plutus Core interpreter has got a lot faster over the last few years (at least one order of magnitude, possibly two by now).
+However, there continues to be relentless pressure on script resource usage.
+This will always be the case: no matter how fast we make things, things can _always_ be faster or cheaper and it will improve throughput and save people money.
+
+That means that significant performance gains are always compelling.
+
+#### 3. Creating and Analyzing Plutus Core Terms is Difficult
+
+As discussed above, both lifting and unlifting are currently very tricky or impossible.
+
+## Specification
+
+This specification has two parts: firstly a change to Plutus Core, and secondly a change to the ledger interface to scripts for a new Plutus language version.
+
+### 1. Plutus Core Changes
+
+The following new term constructors are added to the Plutus Core language:[^typed-plutus-core]
+```
+-- Packs the fields (ts) into a construtor value tagged with i
+(constr i ts)
+
+-- Inspects the tag on t and passes its fields to the corresponding case branch
+(case t cs)
+```
+
+[^typed-plutus-core]: See Appendix 1 for changes to Typed Plutus Core.
+
+The small-step reduction rules for these terms are:
+
+$$
+\mathrm{CONSTR\_SUB}\frac{t_i \rightarrow t'_i}{(\mathrm{constr}\ e\ t_1 \ldots t_n) \rightarrow (\mathrm{constr}\ e\ t'_1 \ldots t'_n)}
+$$
```

Comment:
Yes okay, I guess it should just say you can reduce any one argument.

Reply:
Or possibly I'm confused about what I should be doing here :thinking:

Reply:
Typically you'd distinguish between already evaluated arguments and the remaining redexes and focus in the rule on the first remaining redex, but since we're not trying to please any academics here, we could quantify over `i` right in the name of the rule (thus making the number of rules infinite):

```
                                      t_i ~> t_i'
CONSTR_SUB_i -------------------------------------------------------------
1 <= i <= n  constr e t_1 ... t_i ... t_n ~> constr e t_1 ... t_i' ... t_n
```

This is of course super weird, but

1. who cares
2. it has the big advantage of not specifying the evaluation order, because we probably haven't settled on one yet anyway, but even if we did, it kinda makes sense for this proposal to abstract over such little details

Reply:
Yeah, I did something just like that already.